ntoskrnl/mm/ARM3/expool.c

```diff
@@ -1465,11 +1465,45 @@ ExpExpandBigPageTable(
     ASSERT(KeGetCurrentIrql() == DISPATCH_LEVEL);
 
     /* Make sure we don't overflow */
-    if (!NT_SUCCESS(RtlULongMult(2,
-                                 OldSize * sizeof(POOL_TRACKER_BIG_PAGES),
-                                 &NewSizeInBytes)))
+    if (!Shrink)
+    {
+        if (!NT_SUCCESS(RtlULongMult(2, OldSize, &NewSize)))
+        {
+            DPRINT1("Overflow expanding big page table. Size=%lu\n", OldSize);
+            KeReleaseSpinLock(&ExpLargePoolTableLock, OldIrql);
+            return FALSE;
+        }
+
+        /* Make sure we don't stupidly waste pages */
+        NewSize = ALIGN_DOWN_BY(NewSize, PAGE_SIZE / sizeof(POOL_TRACKER_BIG_PAGES));
+        ASSERT(NewSize > OldSize);
+    }
+    else
     {
-        DPRINT1("Overflow expanding big page table. Size=%lu\n", OldSize);
+        NewSize = OldSize / 2;
+
+        /* Make sure we don't shrink too much. This is not a failure. */
+        if (NewSize < ExpPoolBigEntriesInUse)
```

Comment:
With the only usage inside a condition `if (ExpPoolBigEntriesInUse < (PoolBigPageTableSize / 8))`, this shouldn't happen, right?

Reply:
Hmmm, I remember putting an ASSERT here and triggering it, and this made sense to turn it into a check back then. Let me look again.